zebra-consensus/src/primitives/halo2/tests.rs

```diff
@@ -0,0 +1,240 @@
+
+    let shielded_data: Vec<zebra_chain::orchard::ShieldedData> = (1..=4)
+        .map(|num_recipients| {
+            let mut builder = Builder::new(
+                Flags::from_parts(enable_spends, enable_outputs),
+                Anchor::from_bytes(anchor_bytes).unwrap(),
+            );
+
+            for _ in 0..num_recipients {
+                builder
+                    .add_recipient(None, recipient, NoteValue::from_raw(note_value), None)
+                    .unwrap();
+            }
+
+            let bundle: Bundle<_, i64> = builder.build(rng).unwrap();
+
+            let bundle = bundle
+                .create_proof(&proving_key)
+                .unwrap()
+                .apply_signatures(rng, [0; 32], &[])
+                .unwrap();
+
+            zebra_chain::orchard::ShieldedData {
+                flags,
+                value_balance: note_value.try_into().unwrap(),
+                shared_anchor: anchor_bytes.try_into().unwrap(),
+                proof: zebra_chain::primitives::Halo2Proof(
+                    bundle.authorization().proof().as_ref().into(),
+                ),
+                actions: bundle
+                    .actions()
+                    .iter()
+                    .map(|a| {
+                        let action = zebra_chain::orchard::Action {
+                            cv: a.cv_net().to_bytes().try_into().unwrap(),
+                            nullifier: a.nullifier().to_bytes().try_into().unwrap(),
+                            rk: <[u8; 32]>::from(a.rk()).try_into().unwrap(),
+                            cm_x: pallas::Base::from_bytes(&a.cmx().into()).unwrap(),
+                            ephemeral_key: a.encrypted_note().epk_bytes.try_into().unwrap(),
+                            enc_ciphertext: a.encrypted_note().enc_ciphertext.into(),
+                            out_ciphertext: a.encrypted_note().out_ciphertext.into(),
+                        };
+                        zebra_chain::orchard::shielded_data::AuthorizedAction {
+                            action,
+                            spend_auth_sig: <[u8; 64]>::from(a.authorization()).into(),
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .try_into()
+                    .unwrap(),
+                binding_sig: <[u8; 64]>::from(bundle.authorization().binding_signature())
+                    .try_into()
+                    .unwrap(),
+            }
+        })
+        .collect();
+
+    for sd in shielded_data {
+        println!(
+            "{}",
+            hex::encode(sd.clone().zcash_serialize_to_vec().unwrap())
+        );
+    }
+}
+
+async fn verify_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
+    <V as tower::Service<Item>>::Error: std::fmt::Debug,
```

Comment:
Minor: I thought it might make sense to make it explicit that these are constraints on the same type, to avoid risk of confusion:

```suggestion
    <V as tower::Service<Item>>::Error: From<tower::BoxError> + std::fmt::Debug,
```

Reply:
Done in 4681f761